[CodeCompletion] Completion for `@storageRestrictions` attribute

ahoppen opened the pull request:
Offer 
- `storageRestrictions` after `@`
- `initializes` and `accesses` labels inside `@storageRestrictions`
- Stored properties after the `initializes` and `accesses` labels

rdar://115501026

ahoppen committed:
[CodeCompletion] Add completion for `@storageRestrictions`

ahoppen commented:
@swift-ci Please smoke test

hamishknight commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -1211,45 +1255,61 @@ Parser::parseStorageRestrictionsAttribute(SourceLoc AtLoc, SourceLoc Loc) {
+      if (consumeIf(tok::code_complete)) {
+        if (this->CodeCompletionCallbacks) {
+          this->CodeCompletionCallbacks->completeDeclAttrParam(
+              CustomSyntaxAttributeKind::StorageRestrictions, /*Index*/ 0,
```
To avoid confusion, should this use `StorageRestrictionsCompletionKind::Label` instead of index `0`?

ahoppen replied:
Good catch 👍🏽

ahoppen commented on lib/IDE/CompletionLookup.cpp:
```diff
@@ -2969,6 +2988,44 @@ void CompletionLookup::getAttributeDeclParamCompletions(
     }
     break;
+  case CustomSyntaxAttributeKind::StorageRestrictions:
+    if (!HasLabel) {
```
It doesn’t really matter because we communicate everything via the `ParamIndex`. I just removed the check for `HasLabel`.

ahoppen commented on lib/IDE/CompletionLookup.cpp:
```diff
@@ -2969,6 +2988,44 @@ void CompletionLookup::getAttributeDeclParamCompletions(
+      bool suggestInitializesLabel = false;
+      bool suggestAccessesLabel = false;
+      bool suggestArgument = false;
+      switch (static_cast<StorageRestrictionsCompletionKind>(ParamIndex)) {
```
The question is: How far do we want to add a separate parameter for it? I would say the two options are to either keep it as-is or to do all of the following
- Add a parameter to `getAttributeDeclParamCompletions`
- Add a member to `CodeCompletionCallbacksImpl`, 
- Add a new `completeStorageRestrictionsAttribute` function to `CodeCompletionCallbacksImpl` that takes a `StorageRestrictionsCompletionKind`

I was mostly held back by adding yet another member to `CodeCompletionCallbacksImpl` but if you prefer that, I’m happy to do it. @rintaro Do you have any opinions?

hamishknight replied:
Ah sorry I missed the fact that this is all being stored on the `CodeCompletionCallbacksImpl` 😅 I kind of wonder if we should break this out into a class hierarchy, similar to how we model AST nodes, e.g there should be a base class that stores the completion kind, and then subclasses have their own completion kind with their state, and then there can be further subclasses that store more state, and we can use `dyn_cast` and such when doing the completion lookup. That's a much bigger refactoring though, I'm happy to leave it as this for now.

ahoppen replied:
Good idea. I’ll draft a follow-up PR for it.

ahoppen replied:
OK, I just started doing it and we would need to generate ~300-400 lines of boilerplate just for those class definitions (30 completion kinds, each with a class, member, initializer and `classof` function). While I think it would be nice, it also doesn’t really fit into the current design of `CodeCompletionCallbacksImpl`, which relies a lot on the ad-hoc nature (eg. setting one of the members in a setter or type checking in `doneParsing` if `ParsedExpr` is not a `nullptr`).

We need to revisit how code completion works with ASTGen anyway, so I would prefer to not make any changes right now and instead re-design it in the light of ASTGen.

hamishknight replied:
Fair enough, sounds good!

xedin reviewed (approved):
Thank you for adding this!

ahoppen commented:
@swift-ci Please smoke test

ahoppen committed:
[Parse] Accumulate a parser status while parsing attributes

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@hamishknight @xedin I added one more patch commit to this PR. Could you give it a quick review? https://github.com/apple/swift/pull/68603/commits/712d3e2fe176ce040300c88aedbf3bd7e7611b0f

xedin commented on lib/IDE/CompletionLookup.cpp:
```diff
@@ -3051,6 +3070,43 @@ void CompletionLookup::getAttributeDeclParamCompletions(
+    }
+    if (suggestInitializesLabel) {
+      addDeclAttrParamKeyword("initializes", /*Parameters=*/{},
+                              "Specify variables initialized by the accessor",
```
I think "variables" should be replaced with "stored properties" here and and for "accesses" as well.

xedin commented on test/IDE/complete_storage_restriction.swift:
```diff
@@ -0,0 +1,60 @@
+    get { 1 }
+  }
+  // FIRST_ARGUMENT: Begin completions, 1 item
+  // FIRST_ARGUMENT-DAG: Decl[InstanceVar]/CurrNominal: other[#Int#];
```
Might make sense to add one more test here that checks "accesses:" as well for completeness.

hamishknight reviewed (approved):
LGTM, looks like my comment about `VD->hasStorage()` is still outstanding, but I don't think it's that big of a deal

ahoppen commented on lib/IDE/CompletionLookup.cpp:
```diff
@@ -2166,6 +2167,13 @@ void CompletionLookup::foundDecl(ValueDecl *D, DeclVisibilityKind Reason,
     return;
+  case LookupKind::StoredProperty:
+    if (auto *VD = dyn_cast<VarDecl>(D)) {
+      if (VD->getImplInfo().hasStorage()) {
+        addVarDeclRef(VD, Reason, dynamicLookupInfo);
+      }
```
@xedin I copied this check from https://github.com/apple/swift/blob/cae97a8426910aad471d443c6be23aa88e576dc5/lib/Sema/TypeCheckAttr.cpp#L7323. Do you know/remember why that’s using `VD->getImplInfo().hasStorage()` and not `VD->hasStorage()`?

xedin replied:
`InitAccessorReferencedVariablesRequest` has to be very careful about the requests it triggers to avoid circularity, in this case I wouldn't think that matters match so you probably fine using `VD->hasStorage()` as @hamishknight suggested.

ahoppen replied:
Switched it and uncovered one discrepancy between the two while doing so.

https://github.com/apple/swift/pull/68603/commits/8550799eba4263201c618a2d63b9c4945ce65f93

xedin replied:
It’s always better to use top-level API when possible to avoid surprises :)

ahoppen committed:
[CodeCompletion] Call `hasStorage` instead of `getImplInfo().hasStorage()` to determine if a variable is stored

This uncovered a discrepancy between `getImplInfo().hasStorage()` and `hasStorage()`.

ahoppen commented:
@swift-ci Please smoke test
